packages/client-react-streaming/src/PreloadQuery.tsx

```diff
@@ -0,0 +1,86 @@
+import { SimulatePreloadedQuery } from "./index.cc.js";
+import type {
+  ApolloClient,
+  OperationVariables,
+  QueryOptions,
+  QueryReference,
+} from "@apollo/client";
+import type { ReactNode } from "react";
+import React from "react";
+import { serializeOptions } from "./DataTransportAbstraction/transportedOptions.js";
+import { createTransportedQueryRef } from "./transportedQueryRef.js";
+import type { ProgressEvent } from "./DataTransportAbstraction/DataTransportAbstraction.js";
+
+export type RestrictedPreloadOptions = {
+  fetchPolicy?: "cache-first";
+  returnPartialData?: false;
+  nextFetchPolicy?: undefined;
+  pollInterval?: undefined;
+};
+
+export type PreloadQueryOptions<TVariables, TData> = QueryOptions<
+  TVariables,
+  TData
+> &
+  RestrictedPreloadOptions;
+
+export function PreloadQuery<TData, TVariables extends OperationVariables>({
+  options,
+  getClient,
+  children,
+}: {
+  options: PreloadQueryOptions<TVariables, TData>;
+  getClient: () => ApolloClient<any> | Promise<ApolloClient<any>>;
+  children:
+    | ReactNode
+    | ((
+        queryRef: QueryReference<NoInfer<TData>, NoInfer<TVariables>>
+      ) => ReactNode);
+}) {
+  const preloadOptions = {
+    ...options,
+    fetchPolicy: "cache-first" as const,
+    returnPartialData: false,
+    pollInterval: undefined,
+    nextFetchPolicy: undefined,
+  } satisfies RestrictedPreloadOptions;
+
+  const transportedOptions = sanitizeForTransport(
+    serializeOptions(preloadOptions)
+  );
+
+  const resultPromise = Promise.resolve(getClient())
+    .then((client) => client.query<TData, TVariables>(preloadOptions))
+    .then<Array<Omit<ProgressEvent, "id">>, Array<Omit<ProgressEvent, "id">>>(
+      (result) => [
+        { type: "data", result: sanitizeForTransport(result) },
+        { type: "complete" },
+      ],
+      () => [{ type: "error" }]
+    );
+
+  const queryKey = crypto.randomUUID();
+
+  return (
+    <SimulatePreloadedQuery<TData>
+      options={transportedOptions}
+      result={resultPromise}
+      queryKey={typeof children === "function" ? queryKey : undefined}
+    >
+      {typeof children === "function"
+        ? children(
+            createTransportedQueryRef(
+              transportedOptions,
+              queryKey
+              // we're lying to TypeScript and our users here.
```

Comment:
This makes me wonder, should the transported query ref have a `.toPromise` function on it since its part of the public interface on `QueryReference`? We can always stub it and warn when using it, but might be best to at least conform to the public interface. Thoughts?

Reply:
see https://github.com/apollographql/apollo-client-nextjs/compare/ba35077...e64236d